graphqlbackend: initial on-prem user invites collaborator discovery backend

slimsag opened the pull request:
This is the initial backend for on-prem user invites' collaborator discovery.

Unlike the one we use for Cloud signup, which uses the GitHub API to find collaborators (because the repos are not yet cloned on Sourcegraph), this API leverages gitserver instead since the repositories are cloned. We look at recent committers from 25 of the users' most starred repositories.

This is quite fast, completing in under a few hundred milliseconds.

One drawback here, however, is that we do not get profile pictures currently.
Because these users do not exist on Sourcegraph, and we only have their email
address, we don't have any simple way to get their profile picture. The GitHub API
is not suitable for fetching this information, either, and so instead we fallback to Gravatar
images which are better than nothing.

Signed-off-by: Stephen Gutekanst <stephen@sourcegraph.com>

## Test plan

Unit tests added + manual testing

philipp-spiess commented:
@slimsag [Wiring things up is easy enough](https://github.com/sourcegraph/sourcegraph/pull/31988/commits/b9a58cc031b8f341b176c7e79efd80d6082f974e) however I wasn't able to get the results to give me any results on my dev setup. _I assume_ it is because I also don't manage to be able to connect GitHub and any actual repos to be synced so the git backend is just _empty_ but I’m not sure. Feel free to check out [`ps/on-prem-user-invites-test`](https://github.com/sourcegraph/sourcegraph/pull/31988) which contains your changes rebased onto my UI work and wires the APIs correctly. 🙂

philipp-spiess commented on cmd/frontend/graphqlbackend/user_collaborators.go:
```diff
@@ -0,0 +1,111 @@
+		repo := repo
+		wg.Add(1)
+		goroutine.Go(func() {
+			defer wg.Done()
```
TIL about defer and waitgroups 😄

philipp-spiess commented on cmd/frontend/graphqlbackend/user_collaborators.go:
```diff
@@ -0,0 +1,111 @@
+			_, err := r.db.Users().GetByUsername(ctx, username)
+			return err == nil
+		}
+		_, err := r.db.Users().GetByVerifiedEmail(ctx, email)
+		return err == nil
```
If I understand correctly, [the userExists function will be invoked once with username being an empty string and once with email being an empty string](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/external_service_collaborators.go?L239-247).

However, in our example below we make the username _always_ an empty string so what happens is we run the second part of the query here twice (`r.db.Users().GetByVerifiedEmail(ctx, email)`). Ones with email being an empty string. And that for every result that we find!

I think we should probably guard the second query as well:

```suggestion
		if email != "" {
		  _, err := r.db.Users().GetByVerifiedEmail(ctx, email)
		  return err == nil
		}
		return false
```

slimsag replied:
I've clarified this logic, splitting `userExists` into two variants `userExistsByUsername` and `userExistsByEmail` to make this easier to reason about. I've also guarded against empty usernames effectivey now.

slimsag commented:
Alright, I've added backend tests and also cherry-pick'd your (small) commit to wire up the backend and frontend.

I also added gravatar image URLs, they're not perfect (not everybody has one) but it's better than nothing. See the note inline about why we can't really use GitHub's API to get avatar images otherwise. We can see how that works in practice.

Here's how it looks on my dev instance in dotcom mode, it loads quite snappy:

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/3173176/157355559-c0dcaadb-52ce-4679-a855-f14756b5ad98.png">

philipp-spiess reviewed (approved):
Awesomesauce

limitedmage commented on client/web/src/search/backend.tsx:
```diff
@@ -278,16 +277,27 @@ export function fetchCollaborators(userId: Scalars['ID']): Observable<InvitableC
-    const result = requestGraphQL<ExternalServicesWithCollaboratorsResult, ExternalServicesWithCollaboratorsVariables>(
-        EXTERNAL_SERVICES_WITH_COLLABORATORS,
-        { namespace: userId, first: null, after: null }
+    const result = requestGraphQL<InvitableCollaboratorsResult, InvitableCollaboratorsVariables>(
```
Is it possible to use Apollo instead of `requestGraphQL`?

slimsag replied:
Definitely, I was thinking this too. I'm gonna file an issue for us to follow up on: https://github.com/sourcegraph/sourcegraph/issues/32414
